roofit/roofitcore/inc/UniqueId.h

```diff
@@ -0,0 +1,73 @@
+// Author: Jonas Rembser, CERN, Jun 2021
+
+#ifndef roofit_roofitcore_UniqueId_h
+#define roofit_roofitcore_UniqueId_h
+
+/// A UniqueId can be added as a class member to enhance any class with a
+/// unique identifier for each instantiated object.
+///
+/// Example:
+/// ~~~{.cpp}
+/// class MyClass {
+///
+/// public:
+///    /// Return unique ID by reference.
+///    /// Please always use the name `uniqueId` for the getter.
+///    UniqueId<MyClass> const& uniqueId() const { return _uniqueId; }
+///
+/// private:
+///    const UniqueId<MyClass> _uniqueId; //! should be non-persistent
+///
+/// };
+/// ~~~
+template <class Class>
+struct UniqueId {
+
+public:
+
+    using Value_t = unsigned long;
+
+   /// Create a new UniqueId with the next value from the static counter.
+   UniqueId() : val_{++counter} {}
+
+   // Disable all sorts of copying and moving to ensure uniqueness.
+   UniqueId(const UniqueId&) = delete;
+   UniqueId& operator=(const UniqueId&) = delete;
+   UniqueId(UniqueId&&) = delete;
+   UniqueId& operator=(UniqueId&&) = delete;
+
+   /// Return numerical value of ID.
+   /// Use only if necessary, as the UniqueId type information is lost and
+   /// copying/moving is not prohibited for the value type.
+   /// Please don't turn this into a cast operator, as a function with an
+   /// explicit name is easier to track in the codebase.
+   constexpr Value_t value() const { return val_; }
+
+   bool operator==(UniqueId const &other) const { return val_ == other.val_; }
+   bool operator<(UniqueId const &other) const { return val_ < other.val_; }
+
+   static const UniqueId nullid; /// A value that is less than the ID of any object (similar to nullptr).
+
+private:
+
+   Value_t val_; /// Numerical value of the ID.
```

Comment:
Consider not to switch coding conventions if it's part of RooFitCore?

Reply:
Sorry, that was the muscle memory from CMSSW times....